zebra-consensus/src/block/check.rs

```diff
@@ -169,9 +169,25 @@ pub fn time_is_valid_at(
 ///
 /// `transaction_hashes` is a precomputed list of transaction hashes.
 pub fn merkle_root_validity(
+    network: Network,
     block: &Block,
     transaction_hashes: &[transaction::Hash],
 ) -> Result<(), BlockError> {
+    let nu = NetworkUpgrade::current(network, block.coinbase_height().expect("a valid height"));
+
+    if !block
+        .transactions
+        .iter()
+        .all(|trans| match *trans.as_ref() {
+            transaction::Transaction::V5 {
+                network_upgrade, ..
+            } => network_upgrade == nu,
+            _ => true,
+        })
```

Comment:
Required changes:
* use an exhaustive match - a wildcard match probably has the wrong behaviour for the next network upgrade

Suggestions:
* `use transaction::Transaction`
* use `filter_map` to separate field extraction and field checks
* replace `!all` with `any`

```suggestion
    if block
        .transactions
        .iter()
        .filter_map(|trans| match *trans.as_ref() {
           Transaction::V5 {
                network_upgrade, ..
            } => Some(network_upgrade),
            Transaction::V1 { .. } | Transaction::V2 { .. } | Transaction::V3 { .. } | Transaction::V4 { .. } => None,
        })
        .any(|trans_nu| trans_nu != nu)
```